ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java

```diff
@@ -168,51 +171,80 @@ private int sizeExcludingBlobAndCrcSize() {
+      bufferToSend.putInt(usermetadata.capacity());
+      bufferToSend.put(usermetadata);
+      bufferToSend.putShort((short) blobType.ordinal());
+      short keyLength = blobEncryptionKey == null ? 0 : (short) blobEncryptionKey.remaining();
+      bufferToSend.putShort(keyLength);
+      if (keyLength > 0) {
+        bufferToSend.put(blobEncryptionKey);
+      }
+      bufferToSend.putLong(blobSize);
+      crc.update(bufferToSend.array(), bufferToSend.arrayOffset() + crcStart, bufferToSend.position() - crcStart);
+      nioBuffersFromBlob = blob.nioBuffers();
+      for (ByteBuffer bb : nioBuffersFromBlob) {
+        crc.update(bb);
+        // change it back to 0 since we are going to write it to the channel later.
+        bb.position(0);
+      }
+      crcBuf.putLong(crc.getValue());
+      crcBuf.flip();
+      bufferToSend.flip();
+    }
+  }
+
   @Override
   public long writeTo(WritableByteChannel channel) throws IOException {
     long written = 0;
-    if (sentBytes < sizeInBytes()) {
-      if (bufferToSend == null) {
-        // this is the first time this method was called, prepare the buffer to send the header and other metadata
-        // (everything except the blob content).
-        bufferToSend = ByteBuffer.allocate(sizeExcludingBlobAndCrcSize());
-        writeHeader();
-        int crcStart = bufferToSend.position();
-        bufferToSend.put(blobId.toBytes());
-        BlobPropertiesSerDe.serializeBlobProperties(bufferToSend, properties);
-        bufferToSend.putInt(usermetadata.capacity());
-        bufferToSend.put(usermetadata);
-        bufferToSend.putShort((short) blobType.ordinal());
-        short keyLength = blobEncryptionKey == null ? 0 : (short) blobEncryptionKey.remaining();
-        bufferToSend.putShort(keyLength);
-        if (keyLength > 0) {
-          bufferToSend.put(blobEncryptionKey);
+    try {
+      if (sentBytes < sizeInBytes()) {
+        prepareBuffer();
+        // If the header and metadata are not yet written out completely, try and write out as much of it now.
+        if (bufferToSend.hasRemaining()) {
+          written = channel.write(bufferToSend);
         }
-        bufferToSend.putLong(blobSize);
-        crc.update(bufferToSend.array(), bufferToSend.arrayOffset() + crcStart, bufferToSend.position() - crcStart);
-        crc.update(blob.array(), blob.arrayOffset(), blob.remaining());
-        crcBuf.putLong(crc.getValue());
-        crcBuf.flip();
-        bufferToSend.flip();
-      }
 
-      // If the header and metadata are not yet written out completely, try and write out as much of it now.
-      if (bufferToSend.hasRemaining()) {
-        written = channel.write(bufferToSend);
-      }
+        // If the header and metadata were written out completely (in this call or a previous call),
+        // try and write out as much of the blob now.
+        if (!bufferToSend.hasRemaining() && blob != null) {
+          int n = 0, currentWritten = -1;
```

Comment:
Could you rename `n` to something like `oneTimeBytesWritten`?  This indicates that, when current method is invoked, how many bytes have been written during this call. This may contain one or more nio buffers that are successfully written into channel.

Reply:
will update.